language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -396,81 +411,46 @@ returnCek (FrameUnwrap : ctx) val =
-                        where val' = VBuiltin ex bn arity0 arity (tyargs++[ty]) args argEnv -- reconstruct the bad application
-      TypeArg:arity' ->
-          case arity' of
-            [] -> applyBuiltinName ctx bn args  -- Final argument is a type argument
-            _  -> returnCek ctx $ VBuiltin ex bn arity0 arity' (tyargs++[ty]) args argEnv -- More arguments expected
-instantiateEvaluate _ _ val =
-        throwingWithCause _MachineError NonPolymorphicInstantiationMachineError $ Just val
-
-
--- | Apply a function to an argument and proceed.
--- If the function is a 'LamAbs', then extend the current environment with a new variable and proceed.
--- If the function is a 'Builtin', then check whether we've got the right number of arguments.
--- If we do, then ask the constant application machinery to apply it, and proceed with
--- the result (or throw an error if something goes wrong); if we don't, then add the new
--- argument to the VBuiltin and call returnCek to look for more arguments.
-applyEvaluate
-    :: (GShow uni, GEq uni, DefaultUni <: uni, Closed uni, uni `Everywhere` ExMemoryUsage)
-    => Context uni
-    -> CekValue uni   -- lsh of application
-    -> CekValue uni   -- rhs of application
-    -> CekM uni (Plain Term uni)
-applyEvaluate ctx (VLamAbs _ name _ty body env) arg =
-    computeCek ctx (extendEnv name arg env) body
-applyEvaluate ctx val@(VBuiltin ex bn arity0 arity tyargs args argEnv) arg =do
-    case arity of
-      []        -> throwingWithCause _MachineError EmptyBuiltinArityMachineError $ Just val
-                -- Should be impossible: see instantiateEvaluate.
-      TypeArg:_ -> throwingWithCause _MachineError UnexpectedBuiltinTermArgumentMachineError $ Just val'
-                   where val' = VBuiltin ex bn arity0 arity tyargs (args++[arg]) argEnv -- reconstruct the bad application
-      TermArg:arity' -> do
-          let args' = args ++ [arg]
-          case arity' of
-            [] -> applyBuiltinName ctx bn args' -- 'arg' was the final argument
-            _  -> returnCek ctx $ VBuiltin ex bn arity0 arity' tyargs args' argEnv  -- More arguments expected
-applyEvaluate _ val _ = throwingWithCause _MachineError NonFunctionalApplicationMachineError $ Just val
-
--- | Apply a builtin to a list of CekValue arguments
-applyBuiltinName
+data BArg uni = BTypeArg (Type TyName uni ExMemory) | BTermArg (CekValue uni)
+
+applyBuiltin
+    :: PartialBuiltin uni   -- lsh of application
+    -> BArg uni
+    -> CekM uni (PartialBuiltin uni)
+applyBuiltin pb arg =
+    let val = VBuiltin pb
+    in case (pbRemainingArity pb, arg) of
+        ([],_) -> throwingWithCause _MachineError EmptyBuiltinArityMachineError $ Just val
+        (TypeArg:_, BTermArg _) -> throwingWithCause _MachineError UnexpectedBuiltinTermArgumentMachineError $ Just val
+        (TermArg:_, BTypeArg _) -> throwingWithCause _MachineError UnexpectedBuiltinInstantiationMachineError $ Just val
```

Comment:
Just suggested to throw an error here in the original alternative CEK PR (haven't submitted the full review yet).